crate/operator/update_user_password.py

```diff
@@ -0,0 +1,54 @@
+# CrateDB Kubernetes Operator
+# Copyright (C) 2020 Crate.IO GmbH
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from typing import Any, Dict
+
+from kubernetes_asyncio.client import CoreV1Api
+
+from crate.operator.constants import BACKOFF_TIME
+from crate.operator.cratedb import get_connection, update_user
+from crate.operator.utils.formatting import b64decode
+from crate.operator.utils.kubeapi import get_host, resolve_secret_key_ref
+
+
+async def update_user_password(
+    core: CoreV1Api,
+    namespace: str,
+    name: str,
+    user_spec: Dict[str, Any],
+    old_password: str,
+):
+    """
+    Update the password of a given ``user_spec`` in a CrateDB cluster.
+
+    :param core: An instance of the Kubernetes Core V1 API.
+    :param namespace: The Kubernetes namespace for the CrateDB cluster.
+    :param name: The name for the ``CrateDB`` custom resource. Used to lookup
+        the password for the system user created during deployment.
+    :param user_spec: The user specification from the ``CrateDB`` custom resource
+        users.
+    :param old_password: The old password of the user that should be updated.
+    """
+    host = await get_host(core, namespace, name)
```

Comment:
The way you're calling `update_user_password()`, `get_host()` might be called numerous times for each cluster.

I think a function signature `update_user_password(host, username, old_password, new_password) should be sufficient. You can call `get_host()` once in the `secret_update` handler, and you already have the old and new password in the diff that you're iterating over. You only need to decode both from base64 to plain text.

Reply:
Good point ... makes sense.

The only downside I see is that we have to instantiate `CoreV1Api` before we even know if there is something we have to update. But yes, we save a lot more time if we await the host just once.

Reply:
Yeah, but instantiating `CoreV1Api` is much easier than querying for the host ;-)